### PR TITLE
Trusted proxy handling to existing reverse proxy functionality

### PR DESCRIPTION
**Pull request by jberger**
Both when using self-hosted kubernetes with ingresses and using external cloud proxy services, it is useful to have the ability mark certain proxies as trusted, ignoring their entries in the X-Forwarded-For header and getting the true client's ip address. While there are plugins on CPAN that can do this, those are hard to add to existing pre-built applications which would have to load them, and their mechanism to do so (hooks and rewriting) is heavier that what can be done in core (during message parsing).

This PR adds this ability to core, keeping the implementation as simple as possible and using only core libraries.

### CODE REVIEW

File: lib/Mojo/Message/Request.pm
Comment (kraih) [inline lib/Mojo/Message/Request.pm:308]: One of the examples in the middle should probably be a single address?

File: lib/Mojo/Server.pm
Comment (kraih) [inline lib/Mojo/Server.pm:25]: Looks like this is always true.

File: lib/Mojo/Server.pm
Comment (kraih) [inline lib/Mojo/Server.pm:143]: `... value of the ...`.

File: lib/Mojo/Transaction.pm
Comment (kraih) [inline lib/Mojo/Transaction.pm:255]: `unless the L</"req">` sounds weird, `unless L</"req">` would be similar to how it was before.

File: lib/Mojo/Util.pm
Comment (kraih) [inline lib/Mojo/Util.pm:793]: Think that should be `# True`.

File: lib/Mojo/Util.pm
Comment (kraih) [inline lib/Mojo/Util.pm:797]: `# False`.

File: lib/Mojolicious/Command/daemon.pm
Comment (kraih) [inline lib/Mojolicious/Command/daemon.pm:47]: This is `daemon`, not `prefork`.

File: lib/Mojo/Server.pm
Comment (jberger) [reply lib/Mojo/Server.pm:25]: Yeah, I wondered about that last night. Should I then also check `@$p` or just assign it always?

File: lib/Mojo/Transaction.pm
Comment (jberger) [reply lib/Mojo/Transaction.pm:255]: I was reading it as "unless the request" but as code losing the "the" is probably better. I'm fine with either.

File: lib/Mojolicious/Command/daemon.pm
Comment (jberger) [reply lib/Mojolicious/Command/daemon.pm:47]: good catch, copypasta

File: t/mojo/psgi.t
Comment (kraih) [inline t/mojo/psgi.t:315]: Missing newline.

File: lib/Mojo/Server.pm
Comment (kraih) [reply lib/Mojo/Server.pm:25]: 🤷

Comment (jberger) [reply lib/Mojo/Server.pm:25]: Does `if (my $p = $self->trusted_proxies) { $tx->req->trusted_proxies($p) if @$p }` seem weird? It seems weird to me.

Comment (kraih) [reply lib/Mojo/Server.pm:25]: Very weird.

File: lib/Mojo/Server.pm
Comment (kraih) [inline lib/Mojo/Server.pm:141]: Shouldn't the examples be consistent?

Comment (jberger) [reply lib/Mojo/Server.pm:141]: oh, yes.

Sidenote: i had originally included these 4 original cidrs are they are the private network space defined for these purposes. I'm changing 128.0/8 to 127.0.0.1 because it is probably the most common usage of a private space and I rarely if ever see anyone use anything but .1

File: lib/Mojo/Server.pm
Comment (jberger) [reply lib/Mojo/Server.pm:25]: How about `push @{$tx->req->trusted_proxies}, @{$self->trusted_proxies};`? It has the advantage of shallow cloning (which we probably want anyway) and it avoids lots of tests.

Comment (jberger) [reply lib/Mojo/Server.pm:25]: Also, since this happens after the build tx hook, it is possible they will have set something on the req object.

Comment (jberger) [reply lib/Mojo/Server.pm]: I've gone with this approach.

**Comment by jberger** [comment]
I know we're into the New Year's holiday so I don't expect it to move quickly but I think this PR is ready, I'm calling the vote: @mojolicious/maintainers

File: lib/Mojo/Message/Request.pm
Comment (jhthorsen) [inline lib/Mojo/Message/Request.pm:308]: Could there be a IPv6 address in this example just to make it more obvious v6 supported?

File: lib/Mojo/Util.pm
Comment (jhthorsen) [inline lib/Mojo/Util.pm:187]: I'm not sure if I understand what this function does by looking at the name. Could it be cidr_contains($cidr, $addr), ip_within($addr, $cidr) or something else that implies that it's working with IP addresses?

File: lib/Mojolicious/Command/daemon.pm
Comment (jhthorsen) [inline lib/Mojolicious/Command/daemon.pm:23]: Could we reuse proxy instead? Something like `p|proxy:s{,1}`?

**Review by jhthorsen — COMMENTED** [review]
I'm positive in general to this change. Seems to solve a common problem.

File: lib/Mojo/Util.pm
Comment (kraih) [reply lib/Mojo/Util.pm:187]: I find the name descriptive enough.

Comment (jberger) [reply lib/Mojo/Util.pm:187]: A colleague suggested `cidr_contains` as well and I could be convinced, but I personally think that's too "wonky" of a name. You have to know what it is to know what it is in a way. `network_contains` is similarly accurate and more descriptive to the outside reader imo.

File: lib/Mojolicious/Command/daemon.pm
Comment (jberger) [reply lib/Mojolicious/Command/daemon.pm:23]: Having just reverse proxy mode without trusted ips still has meaning, namely a simple "skip one level". Indeed -t still requires (and implicitly implies) -r because of how the scheme handling works. I can also imagine an extension to -r that would say how many levels to skip, like `-r 2`. All that just goes to say these aren't exactly the same concept and thus I don't want to conflate them at this outermost layer when they aren't internally.

File: lib/Mojo/Message/Request.pm
Comment (jberger) [reply lib/Mojo/Message/Request.pm]: done

**Comment by kraih** [comment]
Like i said on IRC, i'm neutral on this PR.

File: lib/Mojolicious/Command/daemon.pm
Comment (jhthorsen) [reply lib/Mojolicious/Command/daemon.pm:23]: -r or -p..? -p is a boolean flag, which I can’t imagine need to be a supported IP address. That’s why I suggested keeping both. I haven’t tried {,} myself though, so I’m unsure how it works with zero arguments.

File: lib/Mojo/Util.pm
Comment (jhthorsen) [reply lib/Mojo/Util.pm:187]: I suggested ip_within(), since NetAddr::IP has a “within” method. Not a deal breaker for me, but I think it would be nice if we could steal a name from somewhere else.

File: lib/Mojolicious/Command/daemon.pm
Comment (kraih) [reply lib/Mojolicious/Command/daemon.pm:23]: With zero arguments an empty string is passes to the closure.

**Review by marcusramberg — APPROVED** [review]
I think this is a good approach to a common problem 👍

File: lib/Mojolicious/Command/daemon.pm
Comment (jberger) [reply lib/Mojolicious/Command/daemon.pm:23]: Apologies, I meant -p not -r

File: lib/Mojo/Util.pm
Comment (jberger) [reply lib/Mojo/Util.pm:187]: There is no consistency on CPAN. Net::IP calls it "overlaps" with some exported constants to check for different types of overlap, Net::IP::Lite(::Net) calls it "contains" which is where I got my name.

File: lib/Mojolicious/Command/daemon.pm
Comment (kraih) [reply lib/Mojolicious/Command/daemon.pm:23]: Fwiw. I don't see us adding `-p 2` when `-p 127.0.0.1` already works.

File: lib/Mojolicious/Command/daemon.pm
Comment (kraih) [inline lib/Mojolicious/Command/daemon.pm:27]: Pretty sure this can be done with less code too.

**Review by kraih — DISMISSED** [review]
Considering this is a security relevant feature, i think we might need a test in `t/mojolicious/commands.t` that ensures `-p` options end up in the right place.

File: lib/Mojolicious/Command/prefork.pm
Comment (kraih) [inline lib/Mojolicious/Command/prefork.pm:34]: Why not just `$prefork->trusted_proxies([grep {length} @proxy]);`?

File: lib/Mojolicious/Command/prefork.pm
Comment (kraih) [inline lib/Mojolicious/Command/prefork.pm:26]: Something like this would probably work too (with less code):
```
'p|proxy:s' => sub { $_[1] eq '' ? $daemon->reverse_proxy(1) : push @{$daemon->trusted_proxies}, $_[1] },
```

**Comment by kraih** [comment]
This PR has been broken for a few days, so the vote is off for now and we'll consider it a work in progress.

**Comment by jberger** [comment]
I may get some time to work on it again today, but yes, I've considered the vote on hold since you blocked it several days ago

File: t/mojolicious/commands.t
Comment (kraih) [inline t/mojolicious/commands.t:467]: Wrong capitalization (same for the subtests below).

File: lib/Mojolicious/Command/daemon.pm
Comment (kraih) [inline lib/Mojolicious/Command/daemon.pm:34]: `$self->build_server(@args)->run`?

Comment (kraih) [reply lib/Mojolicious/Command/daemon.pm:34]: Or actually:
```perl
sub run { shift->build_server(@_)->run }
```

**Comment by kraih** [comment]
Looks like this is still unfinished, so i'll leave it out of the next release.

File: lib/Mojolicious/Command/daemon.pm
Comment (jhthorsen) [reply lib/Mojolicious/Command/daemon.pm:27]: How about just `$daemon->trusted_proxies([grep {length} @proxy])` ?

File: lib/Mojolicious/Command/prefork.pm
Comment (jberger) [reply lib/Mojolicious/Command/prefork.pm:34]: I guess I was saving a method call if it wasn't needed, but this isn't going to be a hot code path, so it doesn't matter to me

Comment (jberger) [reply lib/Mojolicious/Command/prefork.pm:34]: But I guess that first one handles all of this so, ok

File: lib/Mojolicious/Command/prefork.pm
Comment (jberger) [reply lib/Mojolicious/Command/prefork.pm:26]: this ends up being too long

**Comment by jberger** [comment]
Assuming @kraih doesn't block on security concerns, I'm calling the vote again @mojolicious/maintainers

**Review by kraih — COMMENTED** [review]
I'm neutral on this.

**Review by marcusramberg — DISMISSED** [review]
I wish other web frameworks I have to deal with supported this. Am 👍

File: t/mojolicious/lite_app.t
Comment (kraih) [inline t/mojolicious/lite_app.t:704]: Double newline!

File: lib/Mojo/Util.pm
Comment (kraih) [inline lib/Mojo/Util.pm:203]: `"1"` should be `'1'` i guess.

**Review by marcusramberg — DISMISSED** [review]
I hereby rebless this pr

**Comment by shadowcat-mst** [comment]
LGTM

File: t/mojo/cgi.t
Comment (jhthorsen) [inline t/mojo/cgi.t:194]: No need for quotes. (This is a recurring typo for all the HTTP_X headers)

**Review by shadowcat-mst — APPROVED** [review]
LGTM

File: lib/Mojo/Transaction.pm
Comment (jhthorsen) [inline lib/Mojo/Transaction.pm:47]: I think you can save one line here by doing `return $addr unless...`

File: lib/Mojo/Util.pm
Comment (jhthorsen) [inline lib/Mojo/Util.pm:205]: Can this be golfed into `$addr &= ($mask = pack "B$length", '1' x $mask) if defined $mask`?

**Review by jhthorsen — COMMENTED** [review]
I'm not going to "Request changes", since I think my comments are very minor.

File: lib/Mojo/Util.pm
Comment (jberger) [reply lib/Mojo/Util.pm:205]: wow

File: lib/Mojo/Server.pm
Comment (jberger) [inline lib/Mojo/Server.pm:12]: should this be `sub { $ENV{MOJO_REVERSE_PROXY} || !!@{shift->trusted_proxies} };`

File: lib/Mojo/Util.pm
Comment (jberger) [reply lib/Mojo/Util.pm:205]: Actually I don't even need to keep the updated `$mask` variable so it doesn't need the extra complexity of the assignment inside the parens

**Review by tianon — COMMENTED** [review]
All the tests I see here seem to be testing for mostly sane/valid/benign `X-Forwarded-For` header values -- do you think it would be valuable to include at least one instance of the type of malicious header this is intended to guard against returning malicious values for?

ie, instead of just `10.10.10.10, 192.0.2.2, 192.0.2.1`, perhaps at least one instance of something more malicious like `172.16.6.6, 127.0.0.1, 10.10.10.10, 192.0.2.2, 192.0.2.1`?
(where `10.10.10.10` is the "correct" answer given `127.0/8, 192.0/8`)

**Review by jhthorsen — APPROVED** [review]
LGTM 👍

**Review by marcusramberg — APPROVED** [review]
Third time lucky!

File: lib/Mojolicious/Guides/Cookbook.pod
Comment (tianon) [inline lib/Mojolicious/Guides/Cookbook.pod:112]: I think my own misunderstanding of this PR (which I resolved via discussion on IRC) would've benefited from a concrete example here similar to the ones above this paragraph; essentially explicitly pointing out that `X-Forwarded-For: 7.7.7.7, 192.0.2.2, 192.0.2.1` + `MOJO_TRUSTED_PROXIES='127.0/8, 192.0.2.1'` = `192.0.2.2`.

```pod
In more complex situations, usually involving multiple proxies or proxies that live outside your network, it can be
necessary to tell the application from which ip addresses to expect proxy requests by setting C<MOJO_TRUSTED_PROXIES>
to a list of comma separated addresses or CIDR networks.

For example, setting C<MOJO_TRUSTED_PROXIES> to C<127.0/8, 192.0.2.1> will resolve an C<X-Forwarded-For> header value of
C<7.7.7.7, 192.0.2.2, 192.0.2.1> to C<192.0.2.2>.

For even finer control, L</Rewriting> includes examples of how the changes could be implemented manually.
```

**Pull request merged**
